sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,1512 @@
+    cdef:
+        ITYPE_t n = deref(vecs).size()
+        np.ndarray[object, ndim=1] nd_arrays_of_nd_arrays = np.empty(n,
+                                                                     dtype=np.ndarray)
+
+    for i in range(n):
+        nd_arrays_of_nd_arrays[i] = vector_to_nd_array(&(deref(vecs)[i]))
+
+    return nd_arrays_of_nd_arrays
+
+#####################
+
+cdef class PairwiseDistancesReduction:
+    f"""Abstract class computing a reduction on pairwise
+    distances between a set of vectors (rows) X and another
+    set of vectors (rows) of Y.
+
+    The implementation of the reduction is done parallelized
+    on chunks whose size can be set using ``chunk_size``.
+
+    Parameters
+    ----------
+    datasets_pair: DatasetsPair
+        The pair of dataset to use.
+
+    chunk_size: int, default={CHUNK_SIZE}
+        The number of vectors per chunk.
+
+    n_threads: int, default=-1
+        The number of OpenMP threads to use for the reduction.
+        Parallelism is done on chunks and the sharding of chunks
+        depends on the `strategy` set on :method:`~PairwiseDistancesReduction.compute`.
+
+        -1 means using all processors.
+    """
+
+    cdef:
+        DatasetsPair _datasets_pair
+
+        ITYPE_t n_threads
+        ITYPE_t effective_omp_n_thread
+        ITYPE_t n_samples_chunk, chunk_size
+
+        ITYPE_t n_X, X_n_samples_chunk, X_n_chunks, X_n_samples_remainder
+        ITYPE_t n_Y, Y_n_samples_chunk, Y_n_chunks, Y_n_samples_remainder
+
+    @classmethod
+    def valid_metrics(cls) -> List[str]:
+        excluded = {
+            "pyfunc",  # is relatively slow because we need to coerce data as np arrays
+            "mahalanobis", # is numerically unstable
+            # TODO: In order to support discrete distance metrics, we need to have a
+            # simultaneous sort which breaks ties on indices when distances are identical.
+            # The best might be using a std::sort and a Comparator whic might need
+            # AoS instead of SoA (currently used).
+            "hamming",
+            *BOOL_METRICS,
+        }
+        return sorted({"fast_euclidean", "fast_sqeuclidean",
+                       *METRIC_MAPPING.keys()}.difference(excluded))
+
+    @classmethod
+    def is_usable_for(cls, X, Y, metric) -> bool:
+        # TODO: what's the best coercion for lists?
+        X = np.asarray(X) if isinstance(X, (tuple, list)) else X
+        Y = np.asarray(Y) if isinstance(Y, (tuple, list)) else Y
+        # TODO: support sparse arrays and 32 bits
+        return (not issparse(X) and X.dtype.itemsize == 8 and X.ndim == 2 and
+                not issparse(Y) and Y.dtype.itemsize == 8 and Y.ndim == 2 and
+                metric in cls.valid_metrics())
+
+    @property
+    def datasets_pair(self) ->DatasetsPair:
+        return self._datasets_pair
+
+    def __init__(self,
+        DatasetsPair datasets_pair,
+        ITYPE_t chunk_size=CHUNK_SIZE,
+        n_threads=-1,
+     ):
```

Comment:
Again weird indentation:

```suggestion
    def __init__(
        self,
        DatasetsPair datasets_pair,
        ITYPE_t chunk_size=CHUNK_SIZE,
        n_threads=-1,
    ):
```

This pattern with first arg on the same line is repeated many times in this file.